keps/sig-storage/0035-20190122-inline-csi-volumes-support.md

```diff
@@ -0,0 +1,211 @@
+
+
+## Motivation
+Implementing support for embedding volumes directly in pod specs would allow external CSI drivers to have feature parity with their in-tree counter parts.  There are several reasons why this is desirable:
+* It provides the community a path to move away from in-tree volume plugins to CSI, as designed in a separate proposal https://github.com/kubernetes/community/pull/2199/. 
+* This feature would make it possible to create new types of CSI drivers such as ephemeral volume drivers.  They can be used to inject arbitrary states such as configuration, secrets, identy, variables or similar information inside pods. 
+* This can also provide a migration path for older Flex-style drivers (allowing the deprecation of Flex.)
+
+### Goals 
+* Provide a high level design for inline CSI volumes support
+* Define API changes needed to support this feature
+* Inline CSI volumes should work with existing CSI drivers
+* Inline CSI volumes should support ephemeral-style volumes
+* Ensure that inline CSI volumes usage is secure
+
+### Non-goals
+* This feature does not introduce nor require new CSI spec features
+* This feature should not force existing CSI drivers to change
+
+## User stories
+* As a storage provider, I want to be able to create CSI drivers that support persistent volumes that can be nested within pod specs.  These volumes would work similarly to how my current in-tree drivers work.
+* As a storage provider, I want to use the CSI API to develop arbitrary drivers that can mount ephemeral volumes which follow the lifecycles of pods where they are embedded.   These ephemeral volumes do not follow the full CSI storage workflow (of provision, stage, attach, mount, unmount, detach, unstage, delete)
+* As a user, I want to be able to deploy pods, with CSI volumes embedded in their specs, without the use of PV/PVC.  
+  * An administrator would create the volume ahead of time with a volume handle.  
+  * User can then setup pods with inline CSI volumes using reference of existing handle.
+
+* As a user I want to be able to define pod specs with embedded ephemeral CSI volumes that are created/mounted when the pod is deployed and are deleted when the pod goes away.
+
+### Examples
+
+**Example 1**
+A pod spec with an inline CSI volume.  Note that `volumeHandle` refers to a pre-existing volume.
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: testpod
+spec:
+  containers:
+    ...
+  volumes:
+      - name: vol
+        csi:
+          lifecycle: persistent
+          driver: io.kubernetes.storage.mock
+          volumeAttributes:
+              name: "Mock Volume 1"
+          volumeHandle: "1"
+```
+
+**Example 2**
+A pod spec with an inline CSI volume with ephemeral mode.  Note that when the volume is declared ephemeral, the volumeHandle is not provided.
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: some-pod
+spec:
+  containers:
+    ...
+  volumes:
+      - name: vol
+        csi:
+          lifecycle: ephemeral
```

Comment:
This has a few issues in decreasing levels of importance:
1. The user is in control of it. This means the user can not specify it, give it the same volume handle as another user's ephemeral volume and it would mount their volume. This is a security issue.
2. Because of the restricted api, a given driver may malfunction if switched to ephemeral style when the driver doesn't support it.
3. The user needs to know to specify it.
4. It takes more effort for the user to specify.

Due to all of these, I recommend we change it to a CSIDriver CR flag. Then the driver is flaged as either ephemeral or normal by the driver packager. A driver can not be made to be both normal end ephemeral. We drop the lifecycle flag.

Reply:
@kfox1111 thank you for the input.  I think it escaped me that a driver should either be persistent or ephemeral, but never both.  In that case, the setting can be introduced in the CSIDriver object.

Reply:
lifecycle flag still here. should be removed.

Reply:
should this `lifecycle` field be removed then?

Reply:
ah ok. Fixing.